lib/std/debug.zig

```diff
@@ -27,185 +42,475 @@ pub const runtime_safety = switch (builtin.mode) {
-    mod_info: pdb.ModInfo,
-    module_name: []u8,
-    obj_file_name: []u8,
-
-    populated: bool,
-    symbols: []u8,
-    subsect_info: []u8,
-    checksum_offset: ?usize,
-};
-
 pub const LineInfo = struct {
     line: u64,
     column: u64,
     file_name: []const u8,
     allocator: ?*mem.Allocator,
 
-    fn deinit(self: LineInfo) void {
+    fn deinit(self: @This()) void {
         const allocator = self.allocator orelse return;
         allocator.free(self.file_name);
     }
 };
 
-var stderr_mutex = std.Thread.Mutex{};
+pub const SymbolInfo = struct {
+    symbol_name: []const u8 = "???",
+    compile_unit_name: []const u8 = "???",
+    line_info: ?LineInfo = null,
+
+    fn deinit(self: @This()) void {
+        if (self.line_info) |li| {
+            li.deinit();
+        }
+    }
+};
+
+pub const default_config = struct {
+    /// Get the writer used for `print`, `panicExtra`, and stack trace dumping.
+    pub fn getWriter() std.fs.File.Writer {
+        return std.io.getStdErr().writer();
+    }
+
+    /// Detect the `TTY.Config` for the writer returned by `getWriter`.
+    /// This determines which escape codes can be used.
+    pub fn detectTTYConfig() TTY.Config {
+        var bytes: [128]u8 = undefined;
+        const allocator = &std.heap.FixedBufferAllocator.init(bytes[0..]).allocator;
+        if (process.getEnvVarOwned(allocator, "ZIG_DEBUG_COLOR")) |_| {
+            return .escape_codes;
+        } else |_| {
+            const stderr_file = io.getStdErr();
+            if (stderr_file.supportsAnsiEscapeCodes()) {
+                return .escape_codes;
+            } else if (builtin.os.tag == .windows and stderr_file.isTty()) {
+                return .{ .windows_api = stderr_file };
+            } else {
+                return .no_color;
+            }
+        }
+    }
+
+    /// A struct which maps from addresses to symbols (`SymbolInfo`).
+    pub const SymbolMap = struct {
+        debug_info: DebugInfo,
+
+        pub fn addressToSymbol(self: *@This(), address: usize) !SymbolInfo {
+            const module = self.debug_info.getModuleForAddress(address) catch |err| switch (err) {
+                error.MissingDebugInfo, error.InvalidDebugInfo => {
+                    return SymbolInfo{};
+                },
+                else => return err,
+            };
+
+            return module.addressToSymbol(address);
+        }
+
+        pub fn init(allocator: *std.mem.Allocator) !@This() {
+            return @This(){ .debug_info = try openSelfDebugInfo(allocator) };
+        }
+
+        pub fn deinit(self: *@This()) void {
+            self.debug_info.deinit();
+        }
+    };
+
+    /// Try to write a line from a source file. If the line couldn't be written but
+    /// the error is acceptable (end of file, file not found, etc.), returns false.
+    /// If the line was correctly writen it returns true.
+    pub fn attemptWriteLineFromSourceFile(writer: anytype, line_info: LineInfo) !bool {
```

Comment:
An error _and_ a boolean return value? The `attempt` part in the name makes me expect `!void` as return type. Do we care if it fails successfully? :)

Reply:
Yes. 
Check out https://github.com/rgreenblatt/zig/blob/5200b5fff55c4875cb66819cb1cf26d154663d39/lib/std/debug.zig#L458-L469

TLDR: if the line is actually written, we want to place a cursor on the column. If an unexpected error occurs, we want to be able to bubble that up to the caller (for instance when formatting a stack trace).

`attemptWriteLineFromSourceFile` is a pretty bad name, but I could think of a better one. Better ideas would of course be welcome.